### PR TITLE
Add Hook actionProductPriceCalculation

### LINKED ISSUE #23246: New hook for product price
#### Is your feature request related to a problem?

As a module developer I have a need to set the product price to what I want.

Having a hook `actionGetProductPrice` that could return new price for the product could open quite a few possibilities for modules without using overrides.
* realtime pricing from erp
* totally new discount processes
* etc

#### Describe the solution you'd like

Something like this:

```php
    public function getPrice(
        $tax = true,
        $id_product_attribute = null,
        $decimals = 6,
	$divisor = null,
        $only_reduc = false,
        $usereduc = true,
	$quantity = 1
    ) {
        $origprice = Product::getPriceStatic((int) $this->id, $tax, $id_product_attribute, $decimals, $divisor, $only_reduc, $usereduc, $quantity);
       $hooked_price = Hook::exec('actionGetPrice', ['product'=>$this->id, 'tax'=>$tax, 'id_product_attribute'=>$id_product_attribute, 'decimals'=>$decimals, 'divisor'=>$divisor, 'only_reduc'=>$only_reduc, 'usereduc'=>$usereduc, 'quantity'=>$quantity, 'origprice'=>$origprice];
);
       if($hooked_price) {
          return $hooked_price;
       }

        return $origprice;
    }
```

Totally untested code, just the idea.

Ideally the hook should also have the possible customer and cart objects (if the price of the product depends on other stuff in the cart that can be calculated also).

#### Alternatives you've considered

At the moment override seems to be the only way to affect product pricing in anyway from a module.

#### Additional context

The cart should also take these calculations in to account for this to be usefull.

Another similar attribute is stock level...

### PR DESCRIPTION
**Pull request by carmas123**
Thanks to this hook we can override and than change the price calculation simply intercepting this hook and replacing "price" for price and "specific_price_reduction" for reduction if only_reduc param is true

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Added a new hook to handle price calculation.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23246 
| How to test?      | Edit by @jolelievre : See below
| Possible impacts? | None.
| Related PRs? | https://github.com/PrestaShop/autoupgrade/pull/573

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27927)

### How to test
 You should use the `ps_qualityassurance` module, and hook on the `` hook, you can then use this code

```php
if ((int) $params['id_product'] === 19) {
  if ($params['use_tax']) {
      $params['price'] = 42.00;
  } else {
      $params['price'] = 35.00;
  }
}
```

The effect of this code is that it changes the price of `Customizable Mug` its price is now 35 (tax excluded) and 42 (tax included). You can than see in FO that the price has changed, if you try and buy this product its price will be 42, if you create a cart/order from the BO it will also be 42.

**However**, when you edit the product in the BO you will see the value from DB, which makes sense since the BO form is a representation of the value persisted in DB, this hook simply "overrides" the price value in the checkout process.

### Important note for module developers
It is very important that the hook handles the taxes (but other parameters may be equally important) if you want its price to remain consistent in all situations (including Invoice generation) but the implementation of the hook call is not the responsibility of the core, the module is responsible so it the internal rules break anything it's the module's responsibility.

### CODE REVIEW

File: classes/Product.php
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        Hook::exec('actionProductPriceCalculation', [
```

**Review by Progi1984 — COMMENTED** [review]
A small comment

File: classes/Product.php
Comment (Progi1984) [inline classes/Product.php]:
```suggestion

```

**Review by Progi1984 — COMMENTED** [review]
Could you add this new hook to the `hook.xml` file ?

**Comment by carmas123** [comment]
> Could you add this new hook to the `hook.xml` file ?

@Progi1984 I've done and committed but I need to open a new PR?

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by atomiix** [comment]
@carmas123 thanks for your contribution! Can you open a PR on the [autoupgrade module](https://github.com/PrestaShop/autoupgrade) so old shops can benefits of this new hook when they'll upgrade?
You can have a look at https://github.com/PrestaShop/autoupgrade/pull/468/files#diff-88afe5302c93a3831dbc6c1d07ad862b91f53d2b09e2af0b44f07a91d6eb96faR26 to see how hooks are added ;)

@PrestaShop/qa-functional I think you can try to see if the hook is called when it should by using ps_qualityassurance and registering `actionProductPriceCalculation`

**Comment by carmas123** [comment]
Done
https://github.com/PrestaShop/autoupgrade/pull/475

**Review by AureRita — COMMENTED** [review]
Hi @carmas123 ,

Thank you for your PR, Currently, this hooks seems to be on the FO and on the BO when you create some product, should it be during the creation's process of some products ?  
We can see what it do with the quality assurance module on this video 

https://user-images.githubusercontent.com/100699445/162777558-276d6f52-6b8e-4f5f-95bc-ff095e42f2da.mp4

on the FO we got this 

https://user-images.githubusercontent.com/100699445/162781124-5d2bbe78-7b80-4340-9692-b49f5863d9cb.mp4

**Comment by Progi1984** [comment]
ping @carmas123 for checking feedback from @AureRita

**Comment by kpodemski** [comment]
@AureRita 

Everything on your videos is working correctly, in my opinion.

@carmas123, could you give us feedback? From my point of view, everything here is working as expected.

**Comment by AureRita** [comment]
Hi @kpodemski ,

In my opinion, the result on the FO should be correct, but what we have on the BO, not, Indeed, we have the interraction of this hook before the display of the product page, and not during this display ( like in the FO ) 

waiting for feedback

**Comment by Sasni** [comment]
Can I ask here and ask for help on how to change the price of the product using your modification?  
I modified Product.php and registered the hook. Should I now be able to set a custom price using 
```
public function hookActionProductPriceCalculation($params){ }
```
in my module? 
Unfortunately I don't know how to do it.

**Comment by carmas123** [comment]
> 

If you have changed the `Product.php` file you now can create into your module a custom hook and change the `$params["price"]` value.

Like this:

`$params["price"] = 10;

Until this pull request it's not merged you need to edit Product.php file manually adding the hook call like my merge pull request.

Regards

**Comment by AureRita** [comment]
Hi @carmas123 

Can we have a feedback about this 

> In my opinion, the result on the FO should be correct, but what we have on the BO, not, Indeed, we have the interraction of this hook before the display of the product page, and not during this display ( like in the FO )

is it a waiting behaviour or not ? 

waiting for your feedback

**Comment by Sasni** [comment]
> is it a waiting behaviour or not ?

The idea is great, but I think **it should not work** on BO. If the price is calculated "individually" Based on the custom calculation of the product (of what the customer chooses) / (or what I get from an external API)  . Then how should your fancy module know what price to show on the backend? 
Well, unless we somehow set in advance what the default prices should be....

**Review by Sasni — CHANGES_REQUESTED** [review]
I think your hook should be in a different place, for example, before calculating taxes.  
About 3850 line in orginal file (or 3877 in your file) just **after** the condition : 

`if (!$specific_price || $specific_price['price'] < 0) {
            $price = (float) $result['price'];
        } else {
            $price = (float) $specific_price['price'];
        }`
before this comment
`        // convert only if the specific price is in the default currency (id_currency = 0)`
This will allow you to set, for example, discounts or taxes from the backend without having to write it in an additional module.

**Comment by carmas123** [comment]
I don't know what do you say with "from the backend" this hook is inthented only for use it in a custom module for a customized price handled by a module and not for show it into the BO. At least for now.

**Comment by Sasni** [comment]
> I don't know what do you say with "from the backend" this hook is inthented only for use it in a custom module for a customized price handled by a module and not for show it into the BO. At least for now.

Let's assume that the price of my product is retrieved from an external API based on the select fields, the customer selects the appropriate options. I then get the price without tax. Of course, I can in the module multiply this value by the amount of tax but this is additional work in a custom module. The same thing is built into Prestashop. 
I would also like to be able to set, for example, a temporary discount (promotion) - I can currently do this with a specific price. But your hook placed at the end of the "priceCalculation" function. - makes it impossible to apply such a discount (because the $price variable is later overwritten) - and forces me to produce the relevant code in my module. 
That's why I suggested placing the Hook where the "$price" variable shows up. In the rest of the code, various operations are performed on this variable (such as adding taxes, modifying this price based on attributes, promotions, etc.). If someone doesn't want to modify the price in this way then they won't set it directly in the product, but if someone wants to (like me) then don't limit this option please :) 

Just to be clear, I don't want and don't see the need to show the custom price of the product in BO.

**Comment by Sasni** [comment]
Sorry, but I have one more question, 
Should the hook "actionProductPriceCalculation" be called 12 times? Is this normal behavior (I ask because I really don't know) 
image
image

**Comment by kpodemski** [comment]
@Sasni, yep, this method is being called in your shopping cart in the header, etc.

**Comment by pstylez** [comment]
hi @carmas123 @Sasni 
First thank you for your work, this hook is exactly what I was looking for for an API prices module linked to my ERP.
I tried the hook, by implemented your modified product class in my prestashop (1.7.8.6)
I then made a module that search the price of a product from its reference via a custom API linked to my ERP and write it in the hookActionProductPriceCalculation($params) function. 
It seems to worked fine in the FO but in the BO the TTC prices in the catalog listing are overrided with prices from the module, I will need to add a condition in order to avoid that and it seems weird behaviour.
But more annoying I cannot edit nor add a new product with the updated product class, without debug mode I have a saving message error "unable to save the modification" with debug mode I cannot acces the catalog, I have the error message below. It is a brand new install with classic theme, and when I switch to default product class it works.

err2

**Comment by carmas123** [comment]
> 

Hi @pstylez an thank you.
You can try to extract only my edited part of Procuct class and then add it into source code without replace all Product.php file.

**Comment by pstylez** [comment]
hi @carmas123 ,
it did the job thanks, all good for me 
the hook does exactly what I want for my module, hope the merge will be approved and we will see it native in next updates.
cheers

**Comment by AureRita** [comment]
Hi @carmas123, 

Currently I didn't approve this pull request because the hook is present on the BO, If you give me a reason about it, this pull request can be approved and maybe merged 

Waiting for your feedback 
Thank you

**Comment by carmas123** [comment]
> 

Where is specified BO? I've use category CO and not BO

**Comment by AureRita** [comment]
> Hi @carmas123 ,
> 
> Thank you for your PR, Currently, this hooks seems to be on the FO and on the BO when you create some product, should it be during the creation's process of some products ? We can see what it do with the quality assurance module on this video
> 
> https://user-images.githubusercontent.com/100699445/181004750-bf922648-edaf-445e-b4f7-da3b93f34b22.mp4


>  

As you can see on the video, the hook is present when you clic on a product on the BO in BO> Catalog>Product 

So that's why I ask you the reason of this presence 

Waiting for your feedback

**Comment by matks** [comment]
@carmas123 Friendly reminder

**Comment by carmas123** [comment]
> @carmas123 Friendly reminder

Sorry but after five months I can't understand what I should do.
Now I change the category to CO-BO-FO as requested because this hook concern the product price calculation and "obviously" it interferes with everything related to the products.

I apologize for the delay but I have been very busy during this time.

Thanks to all.

**Comment by matks** [comment]
> > @carmas123 Friendly reminder
> 
> Sorry but after five months I can't understand what I should do. Now I change the category to CO-BO-FO as requested because this hook concern the product price calculation and "obviously" it interferes with everything related to the products.
> 
> I apologize for the delay but I have been very busy during this time.
> 
> Thanks to all.

Thank you for the answer.

It seems there is a big misunderstanding 😄 . I just read the full conversation.

If I got it right: @AureRita checked your PR and has been looking at the impact of adding this hook. He explored on which web page of the software your hook is being executed. His goal is to check this is intended and not an unexpected side-effects that could introduce bugs later.

@AureRita noticed the hook is being executed on the front-office product page, that is expected. But he also noticed when browsing a back-office page the hook is being executed too and he's asking if that is an expected behavior or an unwanted side-effect.

**Comment by carmas123** [comment]
> 

The hook is simply executed because the calling method is executed.

**Comment by Robin-Fischer-PS** [comment]
PIng @AureRita :)

**Comment by AureRita** [comment]
Hi @carmas123 , 

After some talk, I think the hook have it place on the BO, but we have another issue behind this hook, as you can see on this video ; 

https://user-images.githubusercontent.com/100699445/188183196-01c405a7-8b6e-43b0-bdec-5f6342decb10.mp4

maybe there are too many hook calls so it prevents you from being able to access the second page of the shop because of an infinite load

**Comment by carmas123** [comment]
> Hi @carmas123 ,
> 
> After some talk, I think the hook have it place on the BO, but we have another issue behind this hook, as you can see on this video ;
> 
>  Untitled_.Sep.2.2022.5_27.PM.mp4 
> maybe there are too many hook calls so it prevents you from being able to access the second page of the shop because of an infinite load

I think this problem is due to the module "ps_facetedsearch"

**Comment by pstylez** [comment]
@AureRita : hi, I'm using this hook in first itiration of a module I wrote and it works well with pagination (I'm running a fresh install of 1.7.8.6 with classic theme).

**Comment by jolelievre** [comment]
Hi @AureRita 

I'm not sure but I think the problem mainly comes from what you implemented in your hook action. You use `dump` on every call while this hook is supposed to be used to override a price and return the expected value. As you see in your previous example this adds some dump info everywhere on the page and it also generates a warning because you called `dump(Bonjour)` instead of `dump('Bonjour')`

This makes two potential errors in your custom call, besides when the page is changed I guess it is done by ajax, so it probably returns the result in a JSON response, if you dump something in the middle of a JSON response it becomes unparsable thus breaking the pagination system.

TLDR; the bug doesn't come from the hook itself but rather from the implementation you used for your test. You need to implement it correctly.

```php
$parameters['price'] = 42.00;
```

This should force all product to have 42 as their price and should validate the hook behaviour.

@carmas123 correct me if I'm wrong but this is the purpose of this hook right? I see you suggested this in your comments, but it's not always easy to find the relevant info in the list of comments, next time you open this kind of PR I suggest you indicate this kind of explanation directly in the PR description in the `How to test` row, this will make testing much easier and faster for the QA team 😉

**Comment by carmas123** [comment]
Yes it's all right 💪

**Review by jolelievre — DISMISSED** [review]

**Comment by jolelievre** [comment]
@AureRita I just went through your other comments where you mention the hook is also triggered in the BO. The fault is not with the hook but rather with the Core architecture which calls the `Product::priceCalculation` or `Product::getPriceStatic` in so many places that we have absolutely no way of knowing if it's used in a BO or FO context.

**Comment by Lospepes** [comment]
Hello,

First of all, I love this.

I am also looking for a hook to add conditions (not available in specific prices) to my product prices via a module and especially to be able to affect in an unique place, the price excluding taxes of my product. I would like to avoid adding multiples overrides. 

Can this merge be accepted in the future and can we start working with this solution?

I'm asking because I see that the changes have been approved and the ticket still open.

Thanks in advance

**Comment by kpodemski** [comment]
@PrestaShop/qa-team 

Can it be tested by a dev? It's really straightforward to test.

**Comment by jolelievre** [comment]
Hi @Lospepes 

the code has been approved by maintainers, but not validated by tests yet. The QA team and I are working on testing this before it gets merged.

**Review by jolelievre — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @carmas123 

Thanks to the details of @jolelievre (thank you), I tested it and it works, as we can see here :
image

[Untitled_ Nov 4, 2022 2_40 PM.webm](https://user-images.githubusercontent.com/100699445/199986532-65afd357-ccc5-46a3-91b4-cf00067ccaa6.webm)

Thank you for your PR

**Comment by jolelievre** [comment]
Thanks @carmas123 and @AureRita

**Comment by Lospepes** [comment]
Hello,

I'm facing a problem when I use this with my prestashop 1.7.8.7, all is working like @AureRita video but in my invoice pdf the original price (excluding tax) is wrong and it is the good one in the table order_detail => original_product_price.
I update my price only if 'use_reduc' and 'only_reduc' is true.

What I am doing is a bad practice?

I'm probably not in the right place to ask this question and i'm sorry, feel free to redirect me.

Thanks in advance and I can not wait to use this hook with Prestashop 8 !

**Comment by carmas123** [comment]
Hi @Lospepes to handle the correct price calculation product I use this hooks:

```
        $res = $this->registerHook('actionGetProductPropertiesAfter');
        $res &= $this->registerHook('actionGetCartPriceFromCatalogInCart');
        $res &= $this->registerHook('actionProductPriceCalculation');
```

**Comment by bender2405** [comment]
@carmas123 hi, i tried this hook in a module to override prices with api prices, works well for basic products but when a product has combinations it does not update the price when we change a combination it always stay with the overrided default combination price
Any idea on something I might do wrong ?

**Comment by carmas123** [comment]
@bender2405 the best way to handle price calculation is this:

Handle the hook:
```
$this->registerHook('actionProductPriceCalculation');
```

and than, into you module code, use it like this:
```
    public function hookActionProductPriceCalculation($params)
    {
        $this->priceHook($params);
    }

...

    public function priceHook($params)
    {
        $context = $params->context ?? Context::getContext();

        // Find the cart
        $cart = $params["cart"] ?? $context->cart;

        // Cart doesn't specified
        if (!$cart) {
            return;
        }

        // Read decimals
        $decimals = (int)$params["decimals"];

        // Read customer
        $id_customer = (int)$params["id_customer"];

        // Set the product and attribute id
        $id_product = (int)$params["id_product"];
        $id_product_attribute = (int)$params["id_product_attribute"];

        // Read current value
        $value = $params["price"];

        // Recalc the price value!!!
        $params["price"] = calcFinalPrice($context, $value, $decimals, $params, $id_product, $id_product_attribute, $id_customer);
    }
```

**Comment by bender2405** [comment]
@carmas123 okay perfect thanks for your answer now everything works fine except one point :
When I place an order I have my override price all along the steps but when I validate the order the price is not the overrided but the database price, this hook is not used in order validation ?

**Comment by pstylez** [comment]
@carmas123 @bender2405  same issue here on prestashop 8

on product page : overrided price like it's supossed to be
modal cart : database price not the overrided price
regular cart page : overrided price
checkout steps : overrided price
validate checkout, checkout resume : database price and the database price get in the bdd in ps_orders and ps_order_details

I'm not used to price calculation in prestashop but how is it possible, are there different method for price calculation like modal cart need other hook and validate checkout too or am I doing something wrong with the hook, just using this in module with condition controller_type=front so it does not activate in admin pages.

thanks

**Comment by pstylez** [comment]
solve my problem by getting rid of controller=front condition will do other way for not having the hook in admin pages, thanks for this great hook very useful to me.

**Comment by 50l3r** [comment]
This hook exists on `1.7.8.8`? I updated but can't find.

If i install ps_qualityassurance and create a hook `actionProductPriceCalculation` with `var_dump` debug i can't see any changes.

**Pull request merged**
